cirq/protocols/json.py

```diff
@@ -0,0 +1,234 @@
+# Copyright 2019 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+from typing import Union, Any, Dict, Optional, List, Callable, Type, cast, \
+    TYPE_CHECKING
+
+import numpy as np
+import sympy
+from typing_extensions import Protocol
+
+from cirq.type_workarounds import NotImplementedType
+
+if TYPE_CHECKING:
+    import cirq.ops.pauli_gates
+    import cirq.devices.unconstrained_device
+
+
+class _ResolverCache:
+    """Lazily import and build registry to avoid circular imports."""
+
+    def __init__(self):
+        self._crd = None
+
+    @property
+    def cirq_class_resolver_dictionary(self) -> Dict[str, Type]:
+        if self._crd is None:
+            import cirq
+            self._crd = {
+                'CCXPowGate': cirq.CCXPowGate,
+                'CCZPowGate': cirq.CCZPowGate,
```

Comment:
Is it necessary to repeat the names twice? You could have a list of the classes, and then `{a.__name__: a for a in ...}`

Reply:
In that eventuality, we can create the dictionary as `{easy_case} + {hard_stuff}` and have the best of both worlds. This is not a blocking change.